Write content guidelines (#4064)
godotengine/godot-docs: community/contributing/content_guidelines.rst

.. _doc_content_guidelines:

Content guidelines
==================

This document is here to help us assess what we should include in the official
documentation. Below, you will find a couple of principles and recommendations
to write accessible content.

We want to achieve two goals:

> 1. **Empathize with our users**. We should write in a way that makes it easy for

Reviewer: Leaving this here for debate: Should the punctuation be included in the bold markup when the full sentence is bold? I.e.
```suggestion
1. **Empathize with our users.** We should write in a way that makes it easy for
```

I don't know how much of a reference this is but https://www.otago.ac.nz/writing-style/grammar-punctuation/otago635506.html states:

> Use bold sparingly to emphasise text you wish to stand out such as dates, names, deadlines, etc. Punctuation which follows the bold text should not be bold (unless the whole sentence is in bold).

That's what I tend to do myself, so I don't feel very strongly one way or the other.

Reviewer: @NathanLovato I made a suggestion here as an example, but it applies to the whole document (and #4067).
There are more occurrences of complete bold sentences where the punctuation should be bold.

   them to learn from the docs.
2. **Write a complete reference manual**. Our goal here is not to teach
   programming foundations. Instead, we should provide a reference for how

>    Godot’s features work.

Reviewer: I think the de facto convention for apostrophes has been to use `'` instead of the typographical variant that most keyboard layouts can't input `’`.

I did remove some occurrences of `’` in the past, and currently there doesn't seem to be any in the repo.

Guidelines and principles
-------------------------

Below are the guidelines we should strive to follow. They are not hard rules,
though: exceptionally, a topic will require breaking one or more of these.
Still, we should strive to achieve the two goals listed above.

Writing complete and accessible documentation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**A feature doesn’t exist unless it is documented**. If a user can’t find
information about a feature and how it works, it doesn’t exist to them. We
should ensure that we cover everything Godot does.

> .. note::

Reviewer 2: It's a markup tag, it shouldn't be capitalized.

> .. note::

Reviewer 3: Yep, sorry, just came back here to remove that suggestion after realising it was a tag.

   When adding or updating an engine feature, the documentation team needs to

>    know about it. Contributors should open an issue on the godot-docs repository

Reviewer: ```suggestion
   know about it. Contributors should open an issue on the `godot-docs` repository
```

Reviewer: Or inline URL so it's clickable.

>    when their work gets merged and requires documentation.

Reviewer: Nitpick, but for indented sections like that we tend to use 4 spaces, not 3.

>    when their work gets merged and requires documentation.

Reviewer: Indeed. They're also welcome to write documentation themselves (and some do), but since not all devs are good technical writers it's good to have a workflow which enables them to provide technical insights in the new feature for others to write the docs.

That being said this section is about content guidelines, and details on what engine contributors should do to get their features properly documented might be better suited in the engine contribution guidelines (which are being rewritten in #4067, and might be further updated as we're discussing clarifications to the code contribution workflow with core engine devs).

Do your best to keep documents **under 1000 words in length**. If a page goes
past that threshold, consider splitting it into two parts, if possible. Limiting
page size forces us to write concisely and to break large documents so they each
focus on a particular problem.

Make it clear what **problem** each page or section of a page tackles and what
the user will learn from it. Users need to know if they’re reading the correct
guide to solving problems they encounter. For example, instead of writing the

> heading “Signals”, consider writing “Reacting to changes in another object with

Reviewer: Similarly to the use of the typographical single quote, I don't think we should use typographical double quotes in the rst source.

We currently do only in a few pages, probably from your own contributions :)
```
$ rg "“"
tutorials/export/exporting_pcks.rst
58:and go to Project/Export and click on “Export PCK/Zip”. Also make sure to have
94:validation code as a layer on top. The following example expects a “mod.pck”
96:“mod_scene.tscn” test scene in its root.

tutorials/rendering/viewports.rst
190:Godot also provides a way of customizing how everything is drawn inside :ref:`Viewports <class_Viewport>` using “Debug Draw”.

tutorials/scripting/c_sharp/c_sharp_features.rst
99:If you want to connect a signal in the editor, you need to (re)build the project assemblies to see the new signal. This build can be manually triggered by clicking the “Build” button at the top right corner of the editor window.

getting_started/step_by_step/your_first_game.rst
576:Set the ``Playing`` property in the Inspector to “On”.
$ rg "”"
tutorials/export/exporting_pcks.rst
58:and go to Project/Export and click on “Export PCK/Zip”. Also make sure to have
94:validation code as a layer on top. The following example expects a “mod.pck”
96:“mod_scene.tscn” test scene in its root.

tutorials/rendering/viewports.rst
190:Godot also provides a way of customizing how everything is drawn inside :ref:`Viewports <class_Viewport>` using “Debug Draw”.

tutorials/scripting/c_sharp/c_sharp_features.rst
99:If you want to connect a signal in the editor, you need to (re)build the project assemblies to see the new signal. This build can be manually triggered by clicking the “Build” button at the top right corner of the editor window.

getting_started/step_by_step/your_first_game.rst
576:Set the ``Playing`` property in the Inspector to “On”.
```

It's actually a Sphinx **Smart Quotes** feature to convert "plain" quotes `"` and `'` to typographical ones when it builds (and it does it only for text paragraphs and not for inline code or code blocks).

It was disabled recently with #3760 as it creates issues for localization. It could be re-enabled if we agree that all references to Strings with `"some string"` should be made as inline code, and not plain text.

Author: I don't know how these characters ended up here. Might've happened while running the content through Antidote, which I tested for grammar checking at the start of this contract.

> heading “Signals”, consider writing “Reacting to changes in another object with

Reviewer: On that longer section title example, I partially agree, but we should keep in mind that long section titles means long section entries in the side menu, which can make navigation more cumbersome IMO, as you can't parse at a glance what a section covers.

So while being explicit is good, writers should still try to come up with page and section titles which still allow easy visual grepping (e.g. here "Reacting to object changes with signals" would lose a slight part of the meaning but might be a good compromise for less verbosity).

Just raising this as a comment, I'm not necessarily saying that this should be changed or further clarified - but it's something to consider.

signals”. While a bit long, the second title makes it clear what is the purpose
of signals.

If the page assumes specific knowledge of other Godot features, mention it and
link it to the corresponding documentation. For instance, a page about physics
may use signals, in which case we could note that the page that introduces
signals is a pre-requisite.

Limiting cognitive load
~~~~~~~~~~~~~~~~~~~~~~~

Limit the cognitive load required to read the documentation. The simpler and
more explicit language we use, the more efficient it becomes for people to
learn. You can do so by:

1. Introducing only one new concept at a time whenever possible.
2. Using simple English, as we recommend in our writing guidelines.

> 3. Include one or more **concrete usage examples**. Prefer a real-world example

Reviewer: ```suggestion
3. Including one or more **concrete usage examples**. Prefer a real-world example
```
to match "by:" preceding the list.

> 3. Include one or more **concrete usage examples**. Prefer a real-world example
>    to abstract code like ``foobar``.

Reviewer 3: ```suggestion
3. Include one or more **concrete usage examples**. It is preferable to use a 
real-world example as opposed to abstract code concepts like ``foobar``.
```

Reviewer 2: We tend not to use the passive voice, as per [Docs writing guidelines](https://docs.godotengine.org/en/stable/community/contributing/docs_writing_guidelines.html).

Reviewer 3: I'm not very good at avoiding passive voice, however I still wouldn't start the sentence with Prefer. I'd suggest perhaps "Use a real-world example over abstract code like ``foobar``" instead.

Author: In this case, the passive sounds fine, it's a bit softer than using the imperative form.

> While many persons may understand more complex language and abstract examples,

Reviewer 3: ```suggestion
While many people may understand more complex language and abstract examples,
```

you will lose others. Also, understandable writing and practical examples
benefit everyone.

> Always make an effort to **put yourself in the users’ shoes**. When we

Reviewer 3: ```suggestion
Always make an effort to **put yourself in the user's shoes**. When we
```

Reviewer: Both are correct, there's more than one user :)

Google has 10k results for the former, 4k results for the latter (including some "users shoes").

I *guess* we can go with "user's" as there's only one user reading at a time, and you can only fit in so many pairs of shoes at a given time ;) But the plural seems totally fine to me too.

> understand something thoroughly, it becomes evident to us. We may fail to think
> about details relevant to a newcomer. But **good documentation meets users where
> they are**. We should strive to explain each feature’s capabilities or intended

Reviewer 3: ```suggestion
understand something thoroughly, it becomes evident to us. We may fail to think
about details relevant to a newcomer, but **good documentation meets users where
they are**. We should strive to explain each feature’s capabilities or intended
```

uses with the most straightforward language possible.

Try to remember what you first needed to know when learning about the feature or
concept. What new terms did you need to learn? What confused you? What was the
hardest to grasp? You will want users to review your work, and we recommend you

> train to explain the feature before writing about it.

Reviewer 3: I'm not sure what the intention of the word "train" is here, it doesn't really fit.

Reviewer 2: This paragraph seems to talk about preparing yourself for questions and scrutiny and training at explaining the feature you are describing.

Reviewer 3: I still don't think the word "train" is the correct one to use here. "Practice explaining" would be better.

Reviewer: "Practice explaining" sounds good to me.

> hardest to grasp? You will want users to review your work, and we recommend you
> train to explain the feature before writing about it.

Author: ```suggestion
hardest to grasp? You will want users to review your work, and we recommend you
practice explaining the feature before writing about it.
```

> .. note::

Reviewer 3: As above, I didn't spot this as a style convention, but ignore this if it is. I believe this should be capitalised though.
```suggestion
.. Note::
```

>    This principle does not mean we have to assume no prior programming

Reviewer: 4 spaces.

>    This principle does not mean we have to assume no prior programming
>    experience throughout the documentation. Having programming foundations is a

Reviewer 3: ```suggestion
   This principle does not mean we have to assume the reader has no prior programming
   experience throughout the documentation. Having programming foundations is a
```

Reviewer: With that suggestion the sentence because a bit heavy IMO (three verbs), I'd suggest rephrasing if the original is not good as is.

Author: I rephrased and simplified the note

   pre-requisite to use a complex engine like Godot. Talking about variables,

>    functions, or classes is acceptable. But we should favor plain language over

Reviewer 3: ```suggestion
   functions, or classes is acceptable, but we should favor plain language over
```

>    This principle does not mean we have to assume no prior programming
>    experience throughout the documentation. Having programming foundations is a
>    pre-requisite to use a complex engine like Godot. Talking about variables,
>    functions, or classes is acceptable. But we should favor plain language over
>    specific terminology like "metaprogramming". If you need to use precise
>    terms, be sure to define them.

Author: ```suggestion
    Having programming foundations is a pre-requisite to use a complex engine
    like Godot. Talking about variables, functions, or classes is acceptable.
    But we should favor plain language over specific terminology like
    "metaprogramming". If you need to use precise terms, be sure to define them.
```

When a page assumes knowledge of another engine feature, declare it at the

> beginning and link to resources that cover what users need. You may also link to

Reviewer 3: ```suggestion
beginning and link to resources that cover what user's need. You may also link to
```

Reviewer 2: I think you are mistaken here 🙂

Reviewer 3: Sorry, you're correct.

other websites for pre-requisites beyond the documentation’s scope. For example,

> you could link to an intro to programming in the getting started guide or a

Reviewer 3: ```suggestion
you could link to an introduction to programming in the getting started guide, or a
```

> website that teaches math theory in the math section.

Reviewer 3: I know I'm being inconsistent here, but I I would lengthen "intro" to "introduction" while leaving "mathematics" shortened to "math".
